include/cetl/rtti.hpp

```diff
@@ -0,0 +1,316 @@
+    {
+        return _get_static_type_id_();
+    }
+    void* _cast_(const type_id& id) & noexcept override
+    {
+        return (id == _get_static_type_id_()) ? static_cast<void*>(this) : search<FirstBase, OtherBases...>(id);
+    }
+    const void* _cast_(const cetl::type_id& id) const& noexcept override
+    {
+        return (id == _get_static_type_id_()) ? static_cast<const void*>(this) : search<FirstBase, OtherBases...>(id);
+    }
+
+private:
+    // Exhaustively search for a matching conversion throughout the entire type hierarchy tree.
+    // Template parameter pack expansion is not available in C++14 so we do it the hard way.
+    template <typename T>
+    void* search(const type_id& id) noexcept
+    {
+        return T::_cast_(id);
+    }
+    template <typename A, typename B, typename... C>
+    void* search(const type_id& id) noexcept
+    {
+        if (void* const p = search<A>(id))
+        {
+            return p;
+        }
+        return search<B, C...>(id);
+    }
+    // Same but const.
+    template <typename T>
+    const void* search(const type_id& id) const noexcept
+    {
+        return T::_cast_(id);
+    }
+    template <typename A, typename B, typename... C>
+    const void* search(const type_id& id) const noexcept
+    {
+        if (const void* const p = search<A>(id))
+        {
+            return p;
+        }
+        return search<B, C...>(id);
+    }
+};
+
+/// Returns the type ID of the given type.
+/// This function is provided for regularity; it simply forwards the call to \c T::_get_static_type_id_().
+/// The type shall satisfy \ref cetl::has_static_type_id.
+template <typename T>
+CETL_NODISCARD constexpr type_id get_type_id() noexcept
+{
+    return std::decay_t<T>::_get_static_type_id_();
+}
+/// Returns the type ID of the given object.
+/// This overload is for objects that implement the \ref cetl::rtti interface; it simply forwards the result of the
+/// \ref cetl::rtti::_get_polymorphic_type_id_() method.
+template <typename T, std::enable_if_t<has_polymorphic_type_id<T>, int> = 0>
+CETL_NODISCARD type_id get_type_id(const T& obj) noexcept
+{
+    return obj._get_polymorphic_type_id_();
+}
+/// Returns the type ID of the given object.
+/// This overload is selected for objects that do not implement the \ref cetl::rtti interface but satisfy
+/// \ref cetl::has_static_type_id.
+template <typename T, std::enable_if_t<has_static_type_id<T> && !has_polymorphic_type_id<T>, int> = 0>
+CETL_NODISCARD constexpr type_id get_type_id(const T&) noexcept
+{
+    return get_type_id<T>();
+}
+
+/// Performs a safe dynamic type up-/down-conversion in constant time by invoking \ref cetl::rtti::_cast_.
+/// \c T shall be a pointer and \c std::remove_pointer_t<T> shall satisfy \ref cetl::has_static_type_id.
+/// Returns \c nullptr if a safe dynamic type conversion to \c T is not possible.
+/// @{
+template <typename T, typename _from>
+CETL_NODISCARD std::enable_if_t<is_rtti_convertible<_from> &&     //
```

Comment:
Doesn't this require the caller to have access to the concrete type? That sorta defeats the whole purpose.

Reply:
No. It only requires the caller to have access to the target type for the conversion, which may be different from the runtime type (concrete type).

Reply:
My spidey senses are tingling here. I'm not sure it's safe to allow the caller to adjust the pointer (i.e. static_cast) in every circumstance. The queryType method (which is based on techniques that are at least 26 years old) encapsulates any pointer adjustments so the caller doesn't need to reason about them. The compiler will have full visibility into the types involved when it parses the queryType implementation and the only unsafe code is the reinterpret_cast of the returned void* but we know that this pointer is correct and should not be adjusted.

I also don't like the need to extend a template. The whole idea of this type system is that any template usage is pure syntactic sugar. My Janky proposal did not use templates to be a polymorphic type which kept things more POCO-like. The `polymorphic_type_traits` was purely a convenience. The types could be used without any templates if needed.

I'm open to improvements, be they simplifications or worthwhile enhancements, to my IPolymorphicType proposal but I think this current proposal is inferior.

Reply:
Sorry, I still don't understand the objection.

>The compiler will have full visibility into the types involved when it parses the queryType implementation and the only unsafe code is the reinterpret_cast of the returned void* but we know that this pointer is correct and should not be adjusted.

Maybe I'm missing something but this is precisely how my implementation works.

1. The client calls `_cast_` (not directly in my implementation but via a convenience helper, but it's optional).
2. The `_cast_` method is executed on the actual runtime type where a check is done whether the requested type is the same as the runtime type.
  - If the type matches, `static_cast<void*>(this)` is returned. We know that `static_cast<T*>(static_cast<void*>(this))` is safe for any `T` always; the safety of the conversion is not undermined if the inner part is performed inside `_cast_` and the outer part is performed outside of it by the `rtti_cast` helper; hence the whole conversion is safe.
  - If the type does not match, `_cast_` of the first base is invoked; if it returns non-null, its result is returned.
  - Otherwise, the search continues to the next base until the first non-null result or all bases have been traversed.
3. The result is `static_cast` into the requested type. The `_cast_` method has ascertained that the conversion is safe, as shown above.

>I also don't like the need to extend a template. The whole idea of this type system is that any template usage is pure syntactic sugar. My Janky proposal did not use templates to be a polymorphic type, which kept things more POCO-like. The polymorphic_type_traits was purely a convenience. The types could be used without any templates if needed.

I'm not sure about the template either, there are pros and cons. The original design I submitted when I just opened this draft PR did not use the template helper; instead it used a macro to avoid boilerplate. You can inspect it here: https://github.com/OpenCyphal/CETL/pull/90/commits/561ce22477b43c998ef11c1e0366e8606742d68c

Here's the definition of the macro; it is meant to be expanded in the class body:

```c++
/// An optional helper that can be used to implement CETL RTTI support with minimal boilerplate.
/// Use it in the public section of the class definition.
///
/// The first argument is the base class of the class that is being defined, which can be \c cetl::rtti.
///
/// The following arguments are the 16 bytes of the type identifier exposed via \c _get_static_type_id_;
/// if less than 16 bytes are provided, the remaining bytes are zeroed.
#define CETL_RTTI(base, ...)                                                                       \
    static constexpr cetl::type_id _get_static_type_id_() noexcept                                 \
    {                                                                                              \
        return cetl::type_id{__VA_ARGS__};                                                         \
    }                                                                                              \
    cetl::type_id _get_polymorphic_type_id_() const noexcept override                              \
    {                                                                                              \
        return _get_static_type_id_();                                                             \
    }                                                                                              \
    void* _cast_(const cetl::type_id& id)& noexcept override                                       \
    {                                                                                              \
        return (id == _get_static_type_id_()) ? static_cast<void*>(this) : base::_cast_(id);       \
    }                                                                                              \
    const void* _cast_(const cetl::type_id& id) const& noexcept override                           \
    {                                                                                              \
        return (id == _get_static_type_id_()) ? static_cast<const void*>(this) : base::_cast_(id); \
    }
```

This is not a very large macro so one should be able to comfortably survive without it. Its downside is that it doesn't support multiple inheritance; as you can see, it takes only one base, meaning that if T implements more than one interface, it will not be able to cast to a type located in the inheritance branch of the non-first base. To support multiple inheritance, one would need to either use the `rtti_helper<>` base or to traverse the bases manually.

As far as I see, the same considerations are still applicable to your proposal, unless I'm misunderstanding something.

Reply:
Macros are proscribed for anything that isn't conditional includes. CETL takes minimal exceptions for this when not using the polyfill headers (e.g `#if !defined(__cpp_exceptions)`). The polyfill type system must not require the use of macros or all objects using it will require exceptions to the no-macros rule.

The whole reason for this polymorphic type system is to avoid generic programming where this can be considered problematic. The design we come up with here must work without macros nor templates but can provide template helpers. `IPolymorphicType` was a simple C++ interface. We would be able to define and implement polymorphic objects without macros or templates. Use of these is trivial without the template helpers I provided making them pure syntactic sugar. For example:

[Example without any helpers...](https://godbolt.org/z/v6zseEaGn)

[Same example with helpers...](https://godbolt.org/z/YoYcb5asd)

As you can see the helpers only serve to eliminate `reinterpret_cast` from user code and the design is completely useable without them.

Reply:
Finally, `queryType` Allows for true object composition without delegation:

[composition example...](https://godbolt.org/z/baW8jffos)

Reply:
Okay, we seem to be on the same page about the behavioral part of the implementation, as mine does precisely the same thing as yours but with a slightly different API. The composition is also possible as you described:

```c++
class CatDog final : public virtual cetl::rtti
{
  public:
    void* _cast_(const type_id& id) & noexcept override
    {
        if (void* const p = m_cat._cast_(id)) {
            return p;
        }
        return m_dog._cast_(id);
    }

    const void* _cast_(const type_id& id) const& noexcept override
    {
        if (const void* const p = m_cat._cast_(id)) {
            return p;
        }
        return m_dog._cast_(id);
    }

  private:
    Tabby m_cat{};
    Boxer m_dog{};
};
```

The difference is that I put a much stronger emphasis on the use of helpers than you did. I want you to notice that an equally generic implementation of your approach will also require the search method; what you have in the CatDog example is a very special-case implementation of the same search method.

I have identified the following deficiencies in my proposal so far:

- [ ] Use virtual inheritance instead of ordinary inheritance to avoid ambiguities during up-conversions (see the OP post).

- [ ] Update the docs to de-emphasize the use of helpers in favor of manual implementation. Keep the `cetl::rtti_helper<>` in place for the benefit of less template-conscious applications and as a usage example.

- [ ] Make all methods of `cetl::rtti` pure. The `cetl::rtti_helper` will require specialization to properly terminate the search when the top of a branch is reached.

- [ ] Upon a closer look I see that `_get_polymorphic_type_id_` is inessential and can be removed; `_cast_` is sufficient for all uses, including type identity queries (see `cetl::is_instance_of`).